### PR TITLE
Fix #3252: Change logic in adaptNoArgs

### LINKED ISSUE #3252: Implicit numeric conversion
```scala
scala> val i: Long = List(1, 2, 3).sum 
1 |val i: Long = List(1, 2, 3).sum
  |                               ^
  |                               found:    scala.math.Numeric[Int] => Int
  |                               required: Long
```

### PR DESCRIPTION
**Pull request by odersky**
adaptNoArgs would do an implicit argument search only if the result type
conformed to the expected type. #3252 shows that this needs to be revised.

### CODE REVIEW

**Comment by odersky** [comment]
test performance please

**Comment by odersky** [comment]
To review this it's best to look at the first commit in isolation, which is reasonably small. The rest is just moving code around.

File: compiler/src/dotty/tools/dotc/typer/Typer.scala [hunk 1]
```diff
@@ -1994,169 +1994,186 @@ class Typer extends Namer with TypeAssigner with Applications with Implicits wit
...
+    }
+
+    def adaptNoArgs(wtp: Type): Tree = {
+      val ptNorm = underlyingApplied(pt)
+      val functionExpected = defn.isFunctionType(ptNorm)
+      wtp match {
+        case wtp: ExprType =>
+          adaptInterpolated(tree.withType(wtp.resultType), pt)
+        case wtp: ImplicitMethodType
+        if constrainResult(wtp, followAlias(pt)) || !functionExpected =>
```
Comment (allanrenucci) [inline compiler/src/dotty/tools/dotc/typer/Typer.scala]: Does the order of the checks matter here? I see that `constrainResult` generates new constraints if it returns `true`.
If the order does not matter, then it should be less expensive to first check `!functionExpected`.

Comment (odersky) [reply compiler/src/dotty/tools/dotc/typer/Typer.scala]: Good observation. The order does matter, actually. `constrainResult` prunes the constraint if it succeeds so we want to always run it.

**Pull request merged**